File: leetcode/python/1035_uncrossed_lines.py
"""
T: O(N**2)
S: O(N**2)

The question is the same as LCS. Find the maximum result for current node from
the calculation for the previous one. If current node in bottom row does not
match the node in the top row - consider walking through both top and bottom
row.
"""

from typing import List


class Solution:

    def maxUncrossedLines(self, A: List[int], B: List[int]) -> int:
        dp = {}
        a, b = len(A) - 1, len(B) - 1

        def helper(i, j):
            result = 0
            if i > a or j > b:
                return result
            if (i, j) in dp:
                return dp[(i, j)]
            if A[i] == B[j]:
                result = 1 + helper(i + 1, j + 1)
            else:
                result = max(helper(i, j + 1), helper(i + 1, j))
            dp[(i, j)] = result
            return result

        return helper(0, 0)

File: leetcode/java/1035UncrossedLines.java
/**
 * T: O(N**2) S: O(N**2)
 *
 * <p>The question is the same as LCS. Find the maximum result for current node from the calculation
 * for the previous one. If current node in bottom row does not match the node in the top row -
 * consider walking through both top and bottom row.
 */
class Solution {
  public int maxUncrossedLines(int[] A, int[] B) {
    int a = A.length;
    int b = B.length;
    int[][] dp = new int[a + 1][b + 1];
    for (int i = 1; i <= a; i++) {
      for (int j = 1; j <= b; j++) {
        if (A[i - 1] == B[j - 1]) {
          dp[i][j] = dp[i - 1][j - 1] + 1;
        } else {
          dp[i][j] = Math.max(dp[i - 1][j], dp[i][j - 1]);
        }
      }
    }
    return dp[a][b];
  }
}
